File: patterns/_tests/CreationalPatterns/AbstractFactory/cpp/unit.test.cpp
// ============================================================
//  Unit Tests — Abstract Factory: Biome World Builder
//
//  What we test here:
//    · Each concrete terrain satisfies the Terrain contract
//    · Enemy::clone() produces an independent, equal-valued copy
//    · Each factory method returns the correct product type
//    · WorldGenerator calls every factory method exactly once
//    · Swapping a factory completely changes the output
//
//  Compile & run:
//    g++ -std=c++17 -DTESTING \
//        -lgtest -lgtest_main \
//        unit_test.cpp -o unit_test && ./unit_test
//
//  The -DTESTING flag suppresses main() in the source file.
//  See cpp.cpp entry point: #ifndef TESTING ... #endif
// ============================================================

// SECTION:: Include Source
// Pull in all types by including the source directly.
// main() is guarded by #ifndef TESTING in cpp.cpp.
#define TESTING
#include "../../../../CreationalPatterns/AbstractFactory/cpp/cpp.cpp"

#include <gtest/gtest.h>
#include <sstream>

// SECTION:: Terrain Products
TEST(TerrainProducts, SandTerrain_Name) {
    SandTerrain t;
    EXPECT_EQ(t.name(), "Sand Dunes");
}
TEST(TerrainProducts, SandTerrain_PenaltyMentionsSpeed) {
    EXPECT_NE(SandTerrain().movementPenalty().find("-20%"), std::string::npos);
}
TEST(TerrainProducts, SandTerrain_DescriptionNonEmpty) {
    EXPECT_GT(SandTerrain().description().size(), 0u);
}
TEST(TerrainProducts, IceTerrain_Name) {
    EXPECT_EQ(IceTerrain().name(), "Frozen Tundra");
}
TEST(TerrainProducts, IceTerrain_PenaltyMentionsSpeed) {
    EXPECT_NE(IceTerrain().movementPenalty().find("-15%"), std::string::npos);
}
TEST(TerrainProducts, MudTerrain_Name) {
    EXPECT_EQ(MudTerrain().name(), "Dense Forest Floor");
}
TEST(TerrainProducts, MudTerrain_PenaltyMentionsSpeed) {
    EXPECT_NE(MudTerrain().movementPenalty().find("-10%"), std::string::npos);
}

// SECTION:: Enemy Prototype Cloning
TEST(EnemyCloning, Scorpion_Clone_IsDistinctPointer) {
    auto orig  = std::make_shared<Scorpion>(50, 15);
    auto clone = orig->clone();
    EXPECT_NE(orig.get(), clone.get());
}
TEST(EnemyCloning, Scorpion_Clone_SameStats) {
    auto orig  = std::make_shared<Scorpion>(50, 15);
    auto clone = orig->clone();
    EXPECT_EQ(clone->health(), orig->health());
    EXPECT_EQ(clone->damage(), orig->damage());
    EXPECT_EQ(clone->name(),   "Scorpion");
}
TEST(EnemyCloning, PolarBear_Clone_PreservesStats) {
    auto orig  = std::make_shared<PolarBear>(180, 35);
    auto clone = orig->clone();
    EXPECT_NE(orig.get(), clone.get());
    EXPECT_EQ(clone->health(), 180);
    EXPECT_EQ(clone->damage(), 35);
}
TEST(EnemyCloning, Wolf_Clone_IsDistinctPointer) {
    auto orig  = std::make_shared<Wolf>(90, 18);
    auto clone = orig->clone();
    EXPECT_NE(orig.get(), clone.get());
    EXPECT_EQ(clone->health(), orig->health());
}

// SECTION:: Factory Method Isolation
class DesertKitTest : public testing::Test {
protected:
    DesertBiomeKit& kit = DesertBiomeKit::getInstance();
};

TEST_F(DesertKitTest, CreateTerrain_ReturnsSandDunes) {
    EXPECT_EQ(kit.createTerrain()->name(), "Sand Dunes");
}
TEST_F(DesertKitTest, CreateEnemies_ReturnsClones) {
    auto a = kit.createEnemies();
    auto b = kit.createEnemies();
    ASSERT_FALSE(a.empty());
    EXPECT_NE(a[0].get(), b[0].get());
    EXPECT_EQ(a[0]->name(), b[0]->name());
}
TEST_F(DesertKitTest, CreateWeather_BaseIsSandstorm) {
    EXPECT_EQ(kit.createWeather().a.name, "Sandstorm");
}
TEST_F(DesertKitTest, CreateWeather_HasHeatWaveEvent) {
    auto ws = kit.createWeather();
    EXPECT_TRUE(ws.b.count("heat_wave") > 0);
}
TEST_F(DesertKitTest, CreateLoot_RollsItems) {
    EXPECT_GE(kit.createLoot()->roll().size(), 1u);
}
TEST_F(DesertKitTest, RegisterEnemy_ExpandsPool) {
    auto before = kit.createEnemies().size();
    kit.registerEnemy(std::make_shared<Wolf>(90, 18));
    EXPECT_EQ(kit.createEnemies().size(), before + 1);
}

TEST(ArcticKitTest, CreateTerrain_ReturnsFrozenTundra) {
    EXPECT_EQ(ArcticBiomeKit::getInstance().createTerrain()->name(), "Frozen Tundra");
}
TEST(ArcticKitTest, CreateWeather_BaseIsBlizzard) {
    EXPECT_EQ(ArcticBiomeKit::getInstance().createWeather().a.name, "Blizzard");
}

// SECTION:: WorldGenerator Decoupling
TEST(WorldGeneratorTest, CallsEveryFactoryMethodOnce) {
    int terrain = 0, enemies = 0, weather = 0, loot = 0;

    struct SpyKit : BiomeKit {
        int& t; int& e; int& w; int& l;
        SpyKit(int& t, int& e, int& w, int& l) : t(t), e(e), w(w), l(l) {}
        std::string kitName() const override { return "SpyKit"; }
        std::shared_ptr<Terrain> createTerrain() override {
            t++; return std::make_shared<SandTerrain>();
        }
        std::vector<std::shared_ptr<Enemy>> createEnemies() override {
            e++; return {};
        }
        WeatherSystem createWeather() override {
            w++; return WeatherSystem{{"Clear","None"}, {}};
        }
        std::shared_ptr<LootTable> createLoot() override {
            l++;
            struct FakeLoot : LootTable {
                std::string name() const override { return "Mock"; }
                std::vector<std::string> roll() override { return {}; }
            };
            return std::make_shared<FakeLoot>();
        }
        void registerEnemy(std::shared_ptr<Enemy>) override {}
    };

    SpyKit spy{terrain, enemies, weather, loot};
    // Redirect cout to suppress output
    std::ostringstream buf;
    auto* orig = std::cout.rdbuf(buf.rdbuf());
    WorldGenerator(spy).generate();
    std::cout.rdbuf(orig);

    EXPECT_EQ(terrain, 1);
    EXPECT_EQ(enemies, 1);
    EXPECT_EQ(weather, 1);
    EXPECT_EQ(loot,    1);
}

TEST(WorldGeneratorTest, SwappingFactoryChangesTerrain) {
    std::ostringstream desert_buf, arctic_buf;

    auto* orig = std::cout.rdbuf(desert_buf.rdbuf());
    WorldGenerator(DesertBiomeKit::getInstance()).generate();
    std::cout.rdbuf(arctic_buf.rdbuf());
    WorldGenerator(ArcticBiomeKit::getInstance()).generate();
    std::cout.rdbuf(orig);

    EXPECT_NE(desert_buf.str().find("Sand Dunes"),    std::string::npos);
    EXPECT_NE(arctic_buf.str().find("Frozen Tundra"), std::string::npos);
    EXPECT_EQ(desert_buf.str().find("Frozen Tundra"), std::string::npos);
    EXPECT_EQ(arctic_buf.str().find("Sand Dunes"),    std::string::npos);
}

File: patterns/_tests/CreationalPatterns/AbstractFactory/java/unit.test.java
// ============================================================
//  Unit Tests — Abstract Factory: Biome World Builder
//
//  What we test here:
//    · Each concrete terrain satisfies the Terrain contract
//    · Enemy.clone() produces an independent, equal-valued copy
//    · Each factory method returns the correct product type
//    · WorldGenerator calls every factory method exactly once
//    · Swapping a factory completely changes the output
//
//  Compile & run:
//    javac -cp .:<junit-jar> \
//          ../../../../CreationalPatterns/AbstractFactory/java/java.java \
//          UnitTest.java
//    java  -cp .:<junit-jar> org.junit.platform.console.standalone.ConsoleLauncher \
//          --select-class=UnitTest
// ============================================================

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.BeforeEach;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;
import java.io.*;

// SECTION:: Terrain Products
@DisplayName("Terrain Products")
class UnitTest {

    @Nested @DisplayName("Terrain Products")
    class TerrainProducts {
        @Test void sandTerrainName() {
            Terrain t = new SandTerrain();
            assertEquals("Sand Dunes", t.name());
        }
        @Test void sandTerrainPenaltyMentionsSpeed() {
            assertTrue(new SandTerrain().movementPenalty().contains("-20%"));
        }
        @Test void sandTerrainDescriptionNonEmpty() {
            assertFalse(new SandTerrain().description().isEmpty());
        }
        @Test void iceTerrainName() {
            assertEquals("Frozen Tundra", new IceTerrain().name());
        }
        @Test void iceTerrainPenalty() {
            assertTrue(new IceTerrain().movementPenalty().contains("-15%"));
        }
        @Test void mudTerrainName() {
            assertEquals("Dense Forest Floor", new MudTerrain().name());
        }
        @Test void mudTerrainPenalty() {
            assertTrue(new MudTerrain().movementPenalty().contains("-10%"));
        }
    }

    // SECTION:: Enemy Prototype Cloning
    @Nested @DisplayName("Enemy Prototype Cloning")
    class EnemyCloning {
        @Test void scorpionCloneIsDistinctObject() {
            Enemy original = new Scorpion(50, 15);
            Enemy cloned   = original.clone();
            assertNotSame(original, cloned);
        }
        @Test void scorpionCloneHasSameStats() {
            Enemy original = new Scorpion(50, 15);
            Enemy cloned   = original.clone();
            assertEquals(original.health(), cloned.health());
            assertEquals(original.damage(), cloned.damage());
            assertEquals("Scorpion", cloned.name());
        }
        @Test void polarBearClonePreservesStats() {
            Enemy original = new PolarBear(180, 35);
            Enemy cloned   = original.clone();
            assertNotSame(original, cloned);
            assertEquals(180, cloned.health());
            assertEquals(35,  cloned.damage());
        }
        @Test void wolfCloneIsIndependent() {
            // Wolf has no mutable setters; verify clone() returns a new object
            Enemy original = new Wolf(90, 18);
            Enemy cloned   = original.clone();
            assertNotSame(original, cloned);
            assertEquals(original.health(), cloned.health());
        }
    }

    // SECTION:: Factory Method Isolation
    @Nested @DisplayName("Factory Method Isolation — DesertBiomeKit")
    class DesertBiomeKitTests {
        private DesertBiomeKit kit;
        @BeforeEach void setUp() { kit = DesertBiomeKit.getInstance(); }

        @Test void createTerrainReturnsSandDunes() {
            assertEquals("Sand Dunes", kit.createTerrain().name());
        }
        @Test void createEnemiesReturnsClones() {
            List<Enemy> a = kit.createEnemies();
            List<Enemy> b = kit.createEnemies();
            assertNotSame(a.get(0), b.get(0));
            assertEquals(a.get(0).name(), b.get(0).name());
        }
        @Test void createWeatherBaseSandstorm() {
            assertEquals("Sandstorm", kit.createWeather().a().name());
        }
        @Test void createWeatherHasHeatWaveEvent() {
            assertTrue(kit.createWeather().b().containsKey("heat_wave"));
        }
        @Test void createLootRollsItems() {
            assertTrue(kit.createLoot().roll().size() >= 1);
        }
        @Test void registerEnemyExpandsPool() {
            int before = kit.createEnemies().size();
            kit.registerEnemy(new Wolf(90, 18));
            assertEquals(before + 1, kit.createEnemies().size());
        }
    }

    @Nested @DisplayName("Factory Method Isolation — ArcticBiomeKit")
    class ArcticBiomeKitTests {
        @Test void createTerrainReturnsFrozenTundra() {
            assertEquals("Frozen Tundra", ArcticBiomeKit.getInstance().createTerrain().name());
        }
        @Test void createWeatherBaseBlizzard() {
            assertEquals("Blizzard", ArcticBiomeKit.getInstance().createWeather().a().name());
        }
    }

    // SECTION:: WorldGenerator Decoupling
    @Nested @DisplayName("WorldGenerator — BiomeKit Contract")
    class WorldGeneratorTests {
        @Test void callsEveryFactoryMethodOnce() {
            int[] calls = {0, 0, 0, 0}; // terrain, enemies, weather, loot

            BiomeKit spyKit = new BiomeKit() {
                public String   kitName()           { return "SpyKit"; }
                public Terrain  createTerrain()     { calls[0]++; return new SandTerrain(); }
                public List<Enemy> createEnemies()  { calls[1]++; return List.of(); }
                public WeatherSystem createWeather(){ calls[2]++;
                    return new WeatherSystem(new Weather(){
                        public String name(){return "Clear";} public String effect(){return "None";}},
                        Map.of());
                }
                public LootTable createLoot()       { calls[3]++;
                    return new LootTable() {
                        public String name(){return "Mock";}
                        public List<String> roll(){return List.of();}};
                }
                public void registerEnemy(Enemy e)  {}
            };

            new WorldGenerator(spyKit);

            assertEquals(1, calls[0], "createTerrain");
            assertEquals(1, calls[1], "createEnemies");
            assertEquals(1, calls[2], "createWeather");
            assertEquals(1, calls[3], "createLoot");
        }

        @Test void swappingFactoryChangesTerrain() throws Exception {
            PrintStream orig = System.out;
            ByteArrayOutputStream desert = new ByteArrayOutputStream();
            ByteArrayOutputStream arctic  = new ByteArrayOutputStream();

            System.setOut(new PrintStream(desert));
            new WorldGenerator(DesertBiomeKit.getInstance()).generate();
            System.setOut(new PrintStream(arctic));
            new WorldGenerator(ArcticBiomeKit.getInstance()).generate();
            System.setOut(orig);

            assertTrue(desert.toString().contains("Sand Dunes"));
            assertTrue(arctic.toString().contains("Frozen Tundra"));
            assertFalse(desert.toString().contains("Frozen Tundra"));
            assertFalse(arctic.toString().contains("Sand Dunes"));
        }
    }
}
